pkg/addon/cache.go

```diff
@@ -210,17 +257,36 @@ func (u *Cache) discoverAndRefreshRegistry() {
 	u.putRegistry2Cache(registries)
 
 	for _, r := range registries {
-		registryMeta, err := r.ListAddonMeta()
-		if err != nil {
-			log.Logger.Errorf("fail to list registry %s metadata,  %v", r.Name, err)
-			continue
-		}
-		u.putAddonMeta2Cache(r.Name, registryMeta)
-		uiData, err := r.ListUIData(registryMeta, UIMetaOptions)
-		if err != nil {
-			log.Logger.Errorf("fail to get addons from registry %s for cache updating, %v", r.Name, err)
-			continue
+		if !IsVersionRegistry(r) {
+			registryMeta, err := r.ListAddonMeta()
+			if err != nil {
+				log.Logger.Errorf("fail to list registry %s metadata,  %v", r.Name, err)
+				continue
+			}
+			u.putAddonMeta2Cache(r.Name, registryMeta)
+			uiData, err := r.ListUIData(registryMeta, UIMetaOptions)
+			if err != nil {
+				log.Logger.Errorf("fail to get addons from registry %s for cache updating, %v", r.Name, err)
+				continue
+			}
+			u.putAddonUIData2Cache(r.Name, uiData)
+		} else {
+			versionedRegistry := BuildVersionedRegistry(r.Name, r.Helm.URL)
+			uiDatas, err := versionedRegistry.ListAddon()
+			if err != nil {
+				log.Logger.Errorf("fail to get addons from registry %s for cache updating, %v", r.Name, err)
+				continue
+			}
+			for _, addon := range uiDatas {
+				uiData, err := versionedRegistry.GetAddonUIData(context.Background(), addon.Name, addon.Version)
+				if err != nil {
+					log.Logger.Errorf("fail to get addon from registry %s, addon %s version %s for cache updating, %v", addon.Name, r.Name, err)
+					continue
+				}
+				u.putVersionedUIData2Cache(r.Name, addon.Name, addon.Version, uiData)
+				// we also no version key, if use get addonUIData without version will return this vale as latest data.
+				u.putVersionedUIData2Cache(r.Name, addon.Name, "latest", uiData)
```

Comment:
how do you know which one is the latest when there're many versions?

Reply:
listAddon will use `sort.Sort(sort.Reverse(versions))` to make latest addon as the first one.